transports/quic/src/certificate.rs

```diff
@@ -0,0 +1,235 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Certificate handling for libp2p
+//!
+//! This handles generation, signing, and verification.
+//!
+//! This crate uses the `log` crate to emit log output.  Events that will occur normally are output
+//! at `trace` level, while “expected” error conditions (ones that can result during correct use of the
+//! library) are logged at `debug` level.
+use log::{trace, warn};
+const LIBP2P_OID: &[u64] = &[1, 3, 6, 1, 4, 1, 53594, 1, 1];
+const LIBP2P_SIGNING_PREFIX: [u8; 21] = *b"libp2p-tls-handshake:";
+pub const LIBP2P_SIGNING_PREFIX_LENGTH: usize = LIBP2P_SIGNING_PREFIX.len();
+const LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH: usize = 65;
+static LIBP2P_SIGNATURE_ALGORITHM: &rcgen::SignatureAlgorithm = &rcgen::PKCS_ECDSA_P256_SHA256;
+use libp2p_core::identity;
+
+fn encode_signed_key(public_key: identity::PublicKey, signature: &[u8]) -> rcgen::CustomExtension {
+    let public_key = public_key.into_protobuf_encoding();
+    let contents = yasna::construct_der(|writer| {
+        writer.write_sequence(|writer| {
+            writer
+                .next()
+                .write_bitvec_bytes(&public_key, public_key.len() * 8);
+            writer
+                .next()
+                .write_bitvec_bytes(signature, signature.len() * 8);
+        })
+    });
+    rcgen::CustomExtension::from_oid_content(LIBP2P_OID, contents)
+}
+
+fn gen_signed_keypair(keypair: &identity::Keypair) -> (rcgen::KeyPair, rcgen::CustomExtension) {
+    let temp_keypair = rcgen::KeyPair::generate(&LIBP2P_SIGNATURE_ALGORITHM)
+        .expect("we pass valid parameters, and assume we have enough memory and randomness; qed");
+    let mut signing_buf =
+        [0u8; LIBP2P_SIGNING_PREFIX_LENGTH + LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH];
+    let public = temp_keypair.public_key_raw();
+    assert_eq!(
+        public.len(),
+        LIBP2P_SIGNATURE_ALGORITHM_PUBLIC_KEY_LENGTH,
+        "ECDSA public keys are 65 bytes"
+    );
+    signing_buf[..LIBP2P_SIGNING_PREFIX_LENGTH].copy_from_slice(&LIBP2P_SIGNING_PREFIX[..]);
+    signing_buf[LIBP2P_SIGNING_PREFIX_LENGTH..].copy_from_slice(public);
+    let signature = keypair.sign(&signing_buf).expect("signing failed");
+    (
+        temp_keypair,
+        encode_signed_key(keypair.public(), &signature),
+    )
+}
+
+pub fn make_cert(keypair: &identity::Keypair) -> rcgen::Certificate {
```

Comment:
Needs some documentation.

```suggestion
/// Generates a self-signed TLS certificate that includes a libp2p-specific certificate extension
/// containing the public key of the given keypair.
pub fn make_cert(keypair: &identity::Keypair) -> rcgen::Certificate {
```